File: web/src/util/inviteTeam.test.ts
import { describe, expect, it } from "vitest"
import { readFileSync } from "node:fs"
import { fileURLToPath } from "node:url"
import {
  INVITE_DESCRIPTION_SCHEMA,
  INVITE_HASH_HEX_LEN,
  INVITE_TEAM_PREFIX,
  inviteTeamName,
  isInviteTeamSlug,
  marshalInviteDescription,
  normalizeInviteEmail,
  parseInviteDescription,
} from "./inviteTeam"

// schemas/invite-v1.schema.json is the source of truth for the record this
// module hand-mirrors, with no compile-time link between them. Assert the two
// agree so a schema edit that isn't mirrored here (or vice versa) fails CI
// instead of silently drifting — same lockstep guard as submissionTags.test.ts.
// The teacher CLI writes and reads these teams too (cli/shared/contract
// InviteTeamPrefix / InviteHashHexLen / InviteSchemaV1, pinned there by
// contract_test.go). There is no compile-time link, so pin the web half too:
// renaming or resizing on this side alone would leave the CLI matching nothing
// and silently stranding invited emails in the org.
describe("invite team name shape — cross-tool contract", () => {
  it("pins the prefix and hash length the CLI expects", () => {
    expect(INVITE_TEAM_PREFIX).toBe("invite-")
    expect(INVITE_HASH_HEX_LEN).toBe(16)
  })
})

describe("classroom50/invite/v1 — shared vector parity with the teacher CLI", () => {
  // Both the web and `gh teacher roster invite` create these teams and each
  // reads the other's, with no compile-time link between the two writers. These
  // vectors are the shared oracle (the Go suite asserts the same file), so a
  // one-sided change to the hash input or the record bytes fails here: it would
  // make every already-created invite team unlocatable and leave the two
  // writers overwriting each other's description forever.
  const fixtureUrl = new URL(
    "../../../cli/shared/testdata/invite_vectors.json",
    import.meta.url,
  )
  const doc = JSON.parse(readFileSync(fileURLToPath(fixtureUrl), "utf8")) as {
    prefix: string
    hash_hex_len: number
    schema: string
    cases: {
      why: string
      classroom: string
      email: string
      slug: string
      record: string
    }[]
  }

  it("addresses the pinned name shape and schema sentinel", () => {
    expect(doc.prefix).toBe(INVITE_TEAM_PREFIX)
    expect(doc.hash_hex_len).toBe(INVITE_HASH_HEX_LEN)
    expect(doc.schema).toBe(INVITE_DESCRIPTION_SCHEMA)
    expect(doc.cases.length).toBeGreaterThan(0)
  })

  for (const c of doc.cases) {
    it(c.why, async () => {
      expect(await inviteTeamName(c.classroom, c.email)).toBe(c.slug)
      // Exact bytes, not a parsed-equal check: a reconcile compares
      // descriptions for string equality.
      expect(marshalInviteDescription(c)).toBe(c.record)
      expect(parseInviteDescription(c.record)).toEqual({
        schema: INVITE_DESCRIPTION_SCHEMA,
        email: normalizeInviteEmail(c.email),
        classroom: c.classroom,
      })
    })
  }
})

describe("classroom50/invite/v1 — schema/mirror parity", () => {
  const schemaUrl = new URL(
    "../../../schemas/invite-v1.schema.json",
    import.meta.url,
  )
  const schema = JSON.parse(readFileSync(fileURLToPath(schemaUrl), "utf8")) as {
    required: string[]
    properties: Record<string, { const?: string }>
  }

  it("matches the schema sentinel", () => {
    expect(schema.properties.schema.const).toBe(INVITE_DESCRIPTION_SCHEMA)
  })

  it("marshals exactly the schema's property set, and all of its required ones", () => {
    const marshaled = JSON.parse(
      marshalInviteDescription({
        email: "alice@example.com",
        classroom: "cs101",
      }),
    ) as Record<string, unknown>
    const marshaledKeys = Object.keys(marshaled).sort()
    // Every declared property is written, and nothing beyond them.
    expect(marshaledKeys).toEqual(Object.keys(schema.properties).sort())
    // Every required field is present in the encoded record.
    for (const key of schema.required) {
      expect(marshaled[key]).toBeTruthy()
    }
  })

  it("parses back a record that satisfies the schema's required set", () => {
    const record = parseInviteDescription(
      marshalInviteDescription({
        email: "alice@example.com",
        classroom: "cs101",
      }),
    )
    expect(record).not.toBeNull()
    for (const key of schema.required) {
      expect(record).toHaveProperty(key)
    }
  })
})

describe("normalizeInviteEmail", () => {
  it("trims and lowercases", () => {
    expect(normalizeInviteEmail("  Alice@Example.COM ")).toBe(
      "alice@example.com",
    )
  })
})

describe("inviteTeamName", () => {
  it("is deterministic, slug-safe, and 23 chars (invite- + 16 hex)", async () => {
    const name = await inviteTeamName("cs101", "alice@example.com")
    expect(name).toMatch(/^invite-[0-9a-f]{16}$/)
    expect(name).toHaveLength(INVITE_TEAM_PREFIX.length + 16)
    // Slug-safe: GitHub derives the slug unchanged (lowercase, hyphen, no
    // special chars), so name === slug.
    expect(name).toBe(name.toLowerCase())
    const again = await inviteTeamName("cs101", "alice@example.com")
    expect(again).toBe(name)
  })

  it("normalizes the email before hashing (case/whitespace insensitive)", async () => {
    const a = await inviteTeamName("cs101", "alice@example.com")
    const b = await inviteTeamName("cs101", "  ALICE@Example.com  ")
    expect(b).toBe(a)
  })

  it("scopes by classroom: same email in two classrooms -> distinct names", async () => {
    const a = await inviteTeamName("cs101", "alice@example.com")
    const b = await inviteTeamName("cs102", "alice@example.com")
    expect(a).not.toBe(b)
  })

  it("does not collide on the classroom/email boundary", async () => {
    // Without a separator byte, ("ab","c") and ("a","bc") would hash the same
    // input. The \u0000 separator prevents this.
    const a = await inviteTeamName("ab", "c@x")
    const b = await inviteTeamName("a", "bc@x")
    expect(a).not.toBe(b)
  })
})

describe("isInviteTeamSlug", () => {
  it("recognizes invite- teams and rejects others", () => {
    expect(isInviteTeamSlug("invite-0123456789abcdef")).toBe(true)
    expect(isInviteTeamSlug("classroom50-cs101")).toBe(false)
    expect(isInviteTeamSlug("classroom50-cs101-teacher")).toBe(false)
  })
})

describe("parseInviteDescription", () => {
  it("parses a valid v1 record", () => {
    const desc = JSON.stringify({
      schema: INVITE_DESCRIPTION_SCHEMA,
      email: "alice@example.com",
      classroom: "cs101",
    })
    expect(parseInviteDescription(desc)).toEqual({
      schema: INVITE_DESCRIPTION_SCHEMA,
      email: "alice@example.com",
      classroom: "cs101",
    })
  })

  it("requires email and classroom", () => {
    expect(
      parseInviteDescription(
        JSON.stringify({ schema: INVITE_DESCRIPTION_SCHEMA, email: "a@b" }),
      ),
    ).toBeNull()
    expect(
      parseInviteDescription(
        JSON.stringify({ schema: INVITE_DESCRIPTION_SCHEMA, classroom: "cs" }),
      ),
    ).toBeNull()
  })

  it("tolerates unknown fields, including a legacy record's name/section", () => {
    // Additive evolution AND backwards compatibility: an earlier release wrote
    // first_name/last_name/section; the slimmed reader ignores them but still
    // recovers the email.
    const desc = JSON.stringify({
      schema: INVITE_DESCRIPTION_SCHEMA,
      email: "a@b",
      classroom: "cs",
      first_name: "Alice",
      section: "S1",
      futureField: "x",
    })
    const parsed = parseInviteDescription(desc)
    expect(parsed?.email).toBe("a@b")
    expect(parsed?.classroom).toBe("cs")
  })

  it("returns null for wrong schema, plain text, non-JSON, null/empty", () => {
    expect(
      parseInviteDescription(
        JSON.stringify({ schema: "other", email: "a@b", classroom: "cs" }),
      ),
    ).toBeNull()
    expect(parseInviteDescription("just a team")).toBeNull()
    expect(parseInviteDescription("{not json")).toBeNull()
    expect(parseInviteDescription(null)).toBeNull()
    expect(parseInviteDescription(undefined)).toBeNull()
    expect(parseInviteDescription("")).toBeNull()
  })
})

describe("marshalInviteDescription", () => {
  it("encodes exactly schema + email + classroom (PII-minimal)", () => {
    const out = marshalInviteDescription({
      email: "alice@example.com",
      classroom: "cs101",
    })
    expect(JSON.parse(out)).toEqual({
      schema: INVITE_DESCRIPTION_SCHEMA,
      email: "alice@example.com",
      classroom: "cs101",
    })
  })

  it("normalizes the stored email", () => {
    const out = marshalInviteDescription({
      email: "  ALICE@Example.com ",
      classroom: "cs101",
    })
    expect(JSON.parse(out).email).toBe("alice@example.com")
  })

  it("round-trips through parseInviteDescription", () => {
    const out = marshalInviteDescription({
      email: "alice@example.com",
      classroom: "cs101",
    })
    expect(parseInviteDescription(out)).toEqual({
      schema: INVITE_DESCRIPTION_SCHEMA,
      email: "alice@example.com",
      classroom: "cs101",
    })
  })

  it("stays far under GitHub's ~250-char description cap for a long email", () => {
    const out = marshalInviteDescription({
      email: `${"x".repeat(64)}@${"y".repeat(60)}.example.com`,
      classroom: "a-fairly-long-classroom-name",
    })
    expect(out.length).toBeLessThanOrEqual(240)
  })

  it("escapes <, >, & (Go json.Marshal parity)", () => {
    const out = marshalInviteDescription({
      email: "a&b<c>@x",
      classroom: "cs",
    })
    expect(out).toContain("\\u0026")
    expect(out).toContain("\\u003c")
    expect(out).toContain("\\u003e")
    expect(out).not.toMatch(/[<>&]/)
    expect(parseInviteDescription(out)?.email).toBe("a&b<c>@x")
  })

  it("escapes U+2028/U+2029 line/paragraph separators (Go parity)", () => {
    const out = marshalInviteDescription({
      email: "a@b",
      classroom: "cs\u2028x\u2029y",
    })
    expect(out).toContain("\\u2028")
    expect(out).toContain("\\u2029")
    expect(out).not.toMatch(/[\u2028\u2029]/)
    expect(parseInviteDescription(out)?.classroom).toBe("cs\u2028x\u2029y")
  })

  // The other half of the parity contract: Go's json.Marshal agrees with
  // JSON.stringify on EVERY C0 control and DEL, so the escaper must leave them
  // all alone. Escaping \b/\f as \u0008/\u000c "for Go parity" would be the bug.
  // Exhaustive over U+0000–U+001F rather than a sample, because the escaper's
  // claim is that <, >, & and U+2028/U+2029 are the ONLY divergences.
  it("leaves every C0 control and DEL exactly as JSON.stringify writes them (Go parity)", () => {
    // The five controls JSON.stringify gives a short escape; every other C0
    // control takes the lowercase \u00xx form.
    const shortEscapes = new Map([
      [0x08, "\\b"],
      [0x09, "\\t"],
      [0x0a, "\\n"],
      [0x0c, "\\f"],
      [0x0d, "\\r"],
    ])
    for (let cp = 0; cp <= 0x1f; cp++) {
      const classroom = `cs${String.fromCharCode(cp)}x`
      const out = marshalInviteDescription({ email: "a@b", classroom })
      const long = `\\u${cp.toString(16).padStart(4, "0")}`
      const short = shortEscapes.get(cp)
      expect(out, `U+${cp.toString(16)}`).toContain(short ?? long)
      if (short) expect(out, `U+${cp.toString(16)}`).not.toContain(long)
      // An uppercase-hex escape would be a byte difference on its own.
      expect(out).not.toContain(long.toUpperCase())
      expect(parseInviteDescription(out)?.classroom).toBe(classroom)
    }
    // DEL is escaped by neither encoder.
    const del = marshalInviteDescription({
      email: "a@b",
      classroom: "cs\u007fx",
    })
    expect(del).toContain("\u007f")
    expect(del).not.toContain("\\u007f")
  })
})

File: cli/gh-teacher/internal/configrepo/invite_team_test.go
package configrepo

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/foundation50/classroom50-cli-shared/contract"
)

// sharedInviteVectorsPath locates the cross-language golden vectors for the
// invite team name and its description record, also consumed by the web writer's
// parity test (web/src/util/inviteTeam.test.ts).
const sharedInviteVectorsPath = "../../../shared/testdata/invite_vectors.json"

type inviteVectorDoc struct {
	Prefix     string `json:"prefix"`
	HashHexLen int    `json:"hash_hex_len"`
	Schema     string `json:"schema"`
	Cases      []struct {
		Why       string `json:"why"`
		Classroom string `json:"classroom"`
		Email     string `json:"email"`
		Slug      string `json:"slug"`
		Record    string `json:"record"`
	} `json:"cases"`
}

func loadInviteVectors(t *testing.T) inviteVectorDoc {
	t.Helper()
	raw, err := os.ReadFile(filepath.Clean(sharedInviteVectorsPath))
	if err != nil {
		t.Fatalf("read shared fixture: %v", err)
	}
	var doc inviteVectorDoc
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("parse shared fixture: %v", err)
	}
	if len(doc.Cases) == 0 {
		t.Fatal("shared fixture has no cases")
	}
	return doc
}

// TestInviteVectors_SharedParity is the load-bearing cross-writer pin: the CLI
// and the web both create these teams and both read each other's, so a one-sided
// change to the hash input or the record bytes would make every already-created
// invite team unlocatable (and make the two writers overwrite each other's
// description forever). The web suite asserts the same vectors.
func TestInviteVectors_SharedParity(t *testing.T) {
	doc := loadInviteVectors(t)
	if doc.Prefix != contract.InviteTeamPrefix {
		t.Errorf("fixture prefix = %q, want %q", doc.Prefix, contract.InviteTeamPrefix)
	}
	if doc.HashHexLen != contract.InviteHashHexLen {
		t.Errorf("fixture hash_hex_len = %d, want %d", doc.HashHexLen, contract.InviteHashHexLen)
	}
	if doc.Schema != contract.InviteSchemaV1 {
		t.Errorf("fixture schema = %q, want %q", doc.Schema, contract.InviteSchemaV1)
	}

	for _, c := range doc.Cases {
		t.Run(c.Why, func(t *testing.T) {
			if got := InviteTeamName(c.Classroom, c.Email); got != c.Slug {
				t.Errorf("InviteTeamName(%q, %q) = %q, want %q", c.Classroom, c.Email, got, c.Slug)
			}
			// The name doubles as the slug, so the teardown/GC fence must accept
			// every name this writer produces.
			if !IsInviteTeamSlug(c.Slug) {
				t.Errorf("IsInviteTeamSlug(%q) = false; the writer produced a slug the sweep can't match", c.Slug)
			}
			record, err := MarshalInviteDescription(c.Classroom, c.Email)
			if err != nil {
				t.Fatalf("MarshalInviteDescription: %v", err)
			}
			if record != c.Record {
				t.Errorf("MarshalInviteDescription(%q, %q) = %q, want %q (exact bytes: a reconcile compares descriptions for string equality)",
					c.Classroom, c.Email, record, c.Record)
			}
			parsed, ok := ParseInviteDescription(c.Record)
			if !ok {
				t.Fatalf("ParseInviteDescription(%q) = not ok", c.Record)
			}
			if parsed.Email != NormalizeInviteEmail(c.Email) {
				t.Errorf("parsed email = %q, want %q", parsed.Email, NormalizeInviteEmail(c.Email))
			}
			if parsed.Classroom != c.Classroom {
				t.Errorf("parsed classroom = %q, want %q", parsed.Classroom, c.Classroom)
			}
			// The reconcile's trust boundary: the recorded email must hash back
			// to the team it was read from.
			if got := InviteTeamName(parsed.Classroom, parsed.Email); got != c.Slug {
				t.Errorf("record does not re-hash to its team: %q, want %q", got, c.Slug)
			}
		})
	}
}

// TestMarshalInviteDescription_ControlCharParity is the other half of the byte
// contract the shared vectors pin: Go's json.Marshal agrees with JSON.stringify
// on EVERY C0 control and DEL, so the web's Go-parity escaper
// (web/src/util/goJsonEscape.ts) must escape ONLY <, >, &, U+2028 and U+2029.
// Exhaustive over U+0000–U+001F rather than a sample, because the escaper's
// claim is about the whole range: a toolchain that started emitting \u0008 for
// \b — or a long escape for any other control — would silently un-align the two
// writers, and a range checked in part is a claim pinned in part.
func TestMarshalInviteDescription_ControlCharParity(t *testing.T) {
	// The five controls JSON.stringify gives a short escape; every other C0
	// control takes the lowercase \u00xx form, and DEL stays raw.
	shortEscapes := map[rune]string{'\b': `\b`, '\t': `\t`, '\n': `\n`, '\f': `\f`, '\r': `\r`}
	for cp := rune(0); cp <= 0x1F; cp++ {
		out, err := MarshalInviteDescription(fmt.Sprintf("cs%cx", cp), "a@b")
		if err != nil {
			t.Fatalf("MarshalInviteDescription(U+%04X): %v", cp, err)
		}
		want, long := shortEscapes[cp], fmt.Sprintf(`\u%04x`, cp)
		if want == "" {
			want = long
		} else if strings.Contains(out, long) {
			t.Errorf("U+%04X: record %q uses the long escape %q; JSON.stringify emits %q", cp, out, long, want)
		}
		if !strings.Contains(out, want) {
			t.Errorf("U+%04X: record %q is missing %q", cp, out, want)
		}
		// An uppercase-hex escape would be a byte difference on its own.
		if upper := strings.ToUpper(long); upper != long && strings.Contains(out, upper) {
			t.Errorf("U+%04X: record %q uses uppercase hex; JSON.stringify emits lowercase", cp, out)
		}
	}
	out, err := MarshalInviteDescription("cs\u007fx", "a@b")
	if err != nil {
		t.Fatalf("MarshalInviteDescription(DEL): %v", err)
	}
	if !strings.Contains(out, "\u007f") {
		t.Errorf("record %q should carry DEL raw; neither encoder escapes it", out)
	}
}

// A single normalizer feeds both the hash and the stored email, so they can't
// disagree about which address a team belongs to.
func TestNormalizeInviteEmail(t *testing.T) {
	cases := []struct{ in, want string }{
		{"  Alice@Example.COM ", "alice@example.com"},
		{"alice@example.com", "alice@example.com"},
		{"\tA@B\n", "a@b"},
		{"", ""},
	}
	for _, tc := range cases {
		if got := NormalizeInviteEmail(tc.in); got != tc.want {
			t.Errorf("NormalizeInviteEmail(%q) = %q, want %q", tc.in, got, tc.want)
		}
	}
}

// TestInviteTeamName_Shape pins the slug-safety the whole addressing scheme
// rests on: GitHub derives a team's slug from its name, so an `invite-<16 hex>`
// name must come back unchanged for name == slug to hold.
func TestInviteTeamName_Shape(t *testing.T) {
	name := InviteTeamName("cs101", "alice@example.com")
	if want := len(contract.InviteTeamPrefix) + contract.InviteHashHexLen; len(name) != want {
		t.Errorf("len(%q) = %d, want %d", name, len(name), want)
	}
	if name != strings.ToLower(name) {
		t.Errorf("name %q is not lowercase; GitHub would rewrite the slug", name)
	}
	if again := InviteTeamName("cs101", "alice@example.com"); again != name {
		t.Errorf("not deterministic: %q then %q", name, again)
	}
}

// ParseInviteDescription is the reader half, applied to a description the
// accepted invitee can hand-edit: it must never bind a record it can't trust,
// but must tolerate a field a newer writer added.
func TestParseInviteDescription(t *testing.T) {
	valid := `{"schema":"classroom50/invite/v1","email":"a@b","classroom":"cs"}`

	t.Run("tolerates unknown fields", func(t *testing.T) {
		// A legacy release wrote first_name/last_name/section; additive
		// evolution means a newer field must not invalidate the record.
		desc := `{"schema":"classroom50/invite/v1","email":"a@b","classroom":"cs","first_name":"Alice","futureField":42}`
		got, ok := ParseInviteDescription(desc)
		if !ok {
			t.Fatal("ok = false, want a tolerated parse")
		}
		if got.Email != "a@b" || got.Classroom != "cs" {
			t.Errorf("parsed = %+v", got)
		}
	})

	t.Run("round-trips a marshaled record", func(t *testing.T) {
		out, err := MarshalInviteDescription("cs", "  A@B ")
		if err != nil {
			t.Fatalf("MarshalInviteDescription: %v", err)
		}
		got, ok := ParseInviteDescription(out)
		if !ok {
			t.Fatalf("ok = false for %q", out)
		}
		if got.Schema != contract.InviteSchemaV1 || got.Email != "a@b" || got.Classroom != "cs" {
			t.Errorf("parsed = %+v", got)
		}
	})

	for _, tc := range []struct {
		why  string
		desc string
	}{
		{"wrong schema sentinel", `{"schema":"classroom50/invite/v2","email":"a@b","classroom":"cs"}`},
		{"missing schema", `{"email":"a@b","classroom":"cs"}`},
		{"missing email", `{"schema":"classroom50/invite/v1","classroom":"cs"}`},
		{"missing classroom", `{"schema":"classroom50/invite/v1","email":"a@b"}`},
		{"a human-written team description", "Invite only, ask the TA"},
		{"truncated JSON", `{"schema":"classroom50/invite/v1"`},
		{"a JSON array", `["classroom50/invite/v1"]`},
		{"an email present but blank", `{"schema":"classroom50/invite/v1","email":"","classroom":"cs"}`},
		{"empty", ""},
	} {
		t.Run("rejects "+tc.why, func(t *testing.T) {
			if got, ok := ParseInviteDescription(tc.desc); ok {
				t.Errorf("ParseInviteDescription(%q) = %+v, ok; want rejected", tc.desc, got)
			}
		})
	}

	// Sanity: the shared valid form the negative cases are derived from parses.
	if _, ok := ParseInviteDescription(valid); !ok {
		t.Errorf("ParseInviteDescription(%q) = not ok", valid)
	}
}

// The record carries the only PII in this feature and GitHub caps a team
// description at ~250 chars, so an RFC-length address must still fit — there is
// no drop-fields fallback (nothing optional is left to drop).
func TestMarshalInviteDescription_FitsDescriptionCap(t *testing.T) {
	long := strings.Repeat("x", 64) + "@" + strings.Repeat("y", 60) + ".example.com"
	out, err := MarshalInviteDescription("a-fairly-long-classroom-name", long)
	if err != nil {
		t.Fatalf("MarshalInviteDescription: %v", err)
	}
	if len(out) > 240 {
		t.Errorf("record is %d chars, want <= 240 (GitHub's team-description cap is ~250)", len(out))
	}
}
